### PR TITLE
feat: public lottery view, totals data

### PR DESCRIPTION
**Pull request by emilyjablonski**
Related to merge of: https://github.com/bloom-housing/bloom/pull/4275

This PR addresses #4224

- [x] Addresses the issue in full
- [ ] Addresses only certain aspects of the issue

## Description

Adds in the total number of applicants and number of applicants per preference into the public site lottery results page.

I created a new table that stores the totals in a very similar way to how we store lottery positions. I had performance concerns over querying the lottery positions table every time a user needs to see their results page as it will have 1000s of rows after just one lottery is run.

There's also some misc. test coverage.

## How Can This Be Tested/Reviewed?

Submit an application to a listing and claim a preference. Run and release the lottery. Ensure that the total number of applications and number of applications that claimed that preference are correct.

Handling listings without preferences or a user who doesn't claim any is a separate ticket.

## Author Checklist:

- [x] Added QA notes to the issue with applicable URLs
- [x] Reviewed in a desktop view
- [x] Reviewed in a mobile view
- [x] Reviewed considering accessibility
- [x] Added tests covering the changes
- [ ] Made corresponding changes to the documentation
- [ ] Ran `yarn generate:client` and/or created a migration when required

## Review Process:

- Read and understand the issue
- Ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Either (1) explicitly ask a clarifying question, (2) request changes, or (3) approve the PR, even if there are very small remaining changes, if you don't need to re-review after the updates

### CODE REVIEW

File: api/src/services/lottery.service.ts [hunk 1]
```diff
@@ -1113,31 +1132,71 @@ export class LotteryService {
       throw new ForbiddenException();
     }
 
-    const applicationUserId = await this.prisma.applications.findFirstOrThrow({
+    if (!user.userRoles.isAdmin) {
```
Comment (ludtkemorgan) [inline api/src/services/lottery.service.ts:1135]: Public users don't have userRoles attached to them. So when I go through the flow with a public user I created this throws a null pointer exception

File: sites/public/src/pages/account/application/[id]/lottery-results.tsx [hunk 2]
```diff
@@ -39,6 +41,14 @@ export default () => {
                 .publicLotteryResults({ id: applicationId })
                 .then((results) => {
                   setResults(results)
+                  lotteryService
```
Comment (ludtkemorgan) [inline sites/public/src/pages/account/application/[id]/lottery-results.tsx:44]: This network call chain has become pretty intense. Since these calls don't actually rely on the previous data in order to run do we have to chain it like this? Some of these calls could be done simultaneously. For example we have the listing id after fetching the application and don't need to wait for the lottery results to do that call.

File: api/src/services/lottery.service.ts [hunk 3]
```diff
@@ -1113,31 +1132,71 @@ export class LotteryService {
...
+   */
+  public async lotteryTotals(
+    listingId: string,
+    user: User,
+  ): Promise<ApplicationLotteryTotal[]> {
+    if (!user) {
+      throw new ForbiddenException();
+    }
+
+    if (!user.userRoles.isAdmin) {
```
Comment (ludtkemorgan) [inline api/src/services/lottery.service.ts:1182]: Same here about public users

File: sites/public/src/pages/account/application/[id]/lottery-results.tsx (hunk 2 above)
Comment (ludtkemorgan) [reply sites/public/src/pages/account/application/[id]/lottery-results.tsx:44]: Also, we are retrieving more data than we need to. Unless I'm missing something it looks like we are just getting the application in order to get the listing id. I'm wondering if there is some consolidation we can do with the 4 endpoints. Or in other words would it be a viable solution to create one endpoint just for this page? I don't remember where the conversations ended on that discussion

**Review by Thompson1985 — APPROVED** [review]
Thanks guys

**Review by Thompson1985 — COMMENTED** [review]
4224/applicant-number-data

File: api/src/controllers/lottery.controller.ts [hunk 4]
```diff
@@ -163,4 +164,20 @@ export class LotteryController {
       mapTo(User, req['user']),
     );
   }
+
+  @Get(`lotteryTotals/:id`)
```
Comment (YazeedLoonat) [inline api/src/controllers/lottery.controller.ts:168]: nit pick: I'm not a fan of passing id params like this and instead think we should be passing these in as DTOs sorta how the listing controller's `getPaginatedSet` call works 

I dunno if you should change this now but that's the standard I'd really like us to move to so we don't have to remember to add the param validation on a bunch of endpoints

File: api/src/dtos/applications/application-lottery-total.dto.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  @IsDefined({ groups: [ValidationsGroupsEnum.default] })
+  @ApiProperty()
+  listingId: string;
+
+  @Expose()
+  @IsString({ groups: [ValidationsGroupsEnum.default] })
+  @IsUUID(4, { groups: [ValidationsGroupsEnum.default] })
+  @IsDefined({ groups: [ValidationsGroupsEnum.default] })
+  @ApiProperty()
+  multiselectQuestionId: string;
```
Comment (YazeedLoonat) [inline api/src/dtos/applications/application-lottery-total.dto.ts:19]: since this is optional in the prisma schema, should this be optional here?

File: api/src/services/lottery.service.ts [hunk 6]
```diff
@@ -1113,31 +1132,71 @@ export class LotteryService {
...
+      await this.prisma.applications.findFirstOrThrow({
```
Comment (YazeedLoonat) [inline api/src/services/lottery.service.ts:1183]: prisma errors are very hard to parse, we shouldn't rely on their `orThrow()` because troubleshooting them in production will be pretty hard. 

instead I recommend doing what we do in other services where we attempt to grab the record and then do a null check and throw an error we can understand if its missing

**Review by YazeedLoonat — CHANGES_REQUESTED** [review]
looks good! along with what morgan said, had a few throughts as well

File: api/src/controllers/lottery.controller.ts (hunk 4 above)
Comment (Thompson1985) [reply api/src/controllers/lottery.controller.ts:168]: Okay definitely makes sense

File: api/src/dtos/applications/application-lottery-total.dto.ts (hunk 5 above)
Comment (Thompson1985) [reply api/src/dtos/applications/application-lottery-total.dto.ts:19]: Looking into now

File: sites/public/src/pages/account/application/[id]/lottery-results.tsx (hunk 2 above)
Comment (emilyjablonski) [reply sites/public/src/pages/account/application/[id]/lottery-results.tsx:44]: Agreed on the nesting and the endpoint consolidation, that just felt like it could turn into a bigger refactor to undertake here. Having one endpoint here I think also sounds ideal! If that feels blocking for this PR happy to do it now.

File: api/src/dtos/applications/application-lottery-total.dto.ts (hunk 5 above)
Comment (emilyjablonski) [reply api/src/dtos/applications/application-lottery-total.dto.ts:19]: I think so, I was duplicating the behavior with `ApplicationLotteryPosition` - should I change `multiselectQuestionId` to be optional there as well?

File: api/src/controllers/lottery.controller.ts (hunk 4 above)
Comment (emilyjablonski) [reply api/src/controllers/lottery.controller.ts:168]: Sounds good! If that feels blocking here I am happy to update that now.

File: api/src/services/lottery.service.ts (hunk 6 above)
Comment (emilyjablonski) [reply api/src/services/lottery.service.ts:1183]: Gotcha! I was copying patterns from elsewhere, particularly `lotteryStatus` - we have ~30 of instances of `findOrThrow`, should we make a ticket to remove them all?

File: api/src/dtos/applications/application-lottery-total.dto.ts (hunk 5 above)
Comment (YazeedLoonat) [reply api/src/dtos/applications/application-lottery-total.dto.ts:19]: ooh yeah can we change that to be optional as well please!

File: api/src/services/lottery.service.ts (hunk 6 above)
Comment (YazeedLoonat) [reply api/src/services/lottery.service.ts]: If I'm not mistaken all instances of `findOrThrow` are custom functions to encapsulate the logic

if you peep `ami-chart.service.ts ~line 137` you'll notice its a custom function that does a findUnique() and then checks to see

so those should be okay. Its more an issue with the inbuilt orThrow functions in prisma

Comment (emilyjablonski) [reply api/src/services/lottery.service.ts]: Will take a look! If you have space let's peek at this in a huddle

File: api/src/dtos/applications/application-lottery-total.dto.ts (hunk 5 above)
Comment (emilyjablonski) [reply api/src/dtos/applications/application-lottery-total.dto.ts:19]: Can do!

File: sites/partners/src/lib/listings/formTypes.ts [hunk 7]
```diff
@@ -167,6 +167,7 @@ export const formDefaults: FormListing = {
   //     rows: [],
   //   },
   // },
+  applicationLotteryTotals: [],
```
Comment (ludtkemorgan) [inline sites/partners/src/lib/listings/formTypes.ts:170]: nit: unrelated to this PR and I think I introduced it, but can we delete the commented out code above this

File: sites/public/src/pages/account/application/[id]/lottery-results.tsx (hunk 2 above)
Comment (ludtkemorgan) [reply sites/public/src/pages/account/application/[id]/lottery-results.tsx:44]: This is not blocking, but we should probably create a ticket to address it at some point

**Review by ludtkemorgan — APPROVED** [review]
Thanks for making those changes!

**Review by YazeedLoonat — APPROVED** [review]
Looks good! 2 small changes to DTOs
